server/container/guice/cassandra-rabbitmq-guice/src/test/java/org/apache/james/Pop3ServerContract.java

```diff
@@ -0,0 +1,559 @@
+            .addDomain(DOMAIN)
+            .addUser(USER, PASSWORD);
+
+        // Given a message is received
+        SMTPMessageSender smtpMessageSender = new SMTPMessageSender(JMAPTestingConstants.DOMAIN);
+        smtpMessageSender.connect("127.0.0.1", server.getProbe(SmtpGuiceProbe.class).getSmtpPort());
+        smtpMessageSender.sendMessageWithHeaders("bob@" + JMAPTestingConstants.DOMAIN, USER,
+            ClassLoaderUtils.getSystemResourceAsString("attachment.eml"));
+        Thread.sleep(100);
+        Awaitility.await().untilAsserted(() -> assertThat(server.getProbe(SpoolerProbe.class).processingFinished()).isTrue());
+
+        // When I delete it in POP3
+        POP3Client pop3Client = new POP3Client();
+        pop3Client.connect("127.0.0.1", server.getProbe(Pop3GuiceProbe.class).getPop3Port());
+        pop3Client.login(USER, PASSWORD);
+        List<POP3MessageInfo> pop3MessageInfos = ImmutableList.copyOf(pop3Client.listMessages());
+        assertThat(pop3MessageInfos).hasSize(1);
+        pop3Client.deleteMessage(pop3MessageInfos.get(0).number);
+
+        // Then RSET should cancen the delete
+        assertThat(ImmutableList.copyOf(pop3Client.listMessages())).isEmpty();
+        pop3Client.disconnect();
+    }
+
+    @Test
+    default void pendingDeletesAreNotSeenByOtherSessions(GuiceJamesServer server) throws Exception {
+        server.getProbe(DataProbeImpl.class).fluent()
+            .addDomain(DOMAIN)
+            .addUser(USER, PASSWORD);
+
+        // Given a message is received
+        SMTPMessageSender smtpMessageSender = new SMTPMessageSender(JMAPTestingConstants.DOMAIN);
+        smtpMessageSender.connect("127.0.0.1", server.getProbe(SmtpGuiceProbe.class).getSmtpPort());
+        smtpMessageSender.sendMessageWithHeaders("bob@" + JMAPTestingConstants.DOMAIN, USER,
+            ClassLoaderUtils.getSystemResourceAsString("attachment.eml"));
+        Thread.sleep(100);
+        Awaitility.await().untilAsserted(() -> assertThat(server.getProbe(SpoolerProbe.class).processingFinished()).isTrue());
+
+        // When I delete it in POP3
+        POP3Client pop3Client = new POP3Client();
+        pop3Client.connect("127.0.0.1", server.getProbe(Pop3GuiceProbe.class).getPop3Port());
+        pop3Client.login(USER, PASSWORD);
+        List<POP3MessageInfo> pop3MessageInfos = ImmutableList.copyOf(pop3Client.listMessages());
+        assertThat(pop3MessageInfos).hasSize(1);
+        pop3Client.deleteMessage(pop3MessageInfos.get(0).number);
+
+        // Then other sessions can still retrieve them
+        POP3Client pop3Client2 = new POP3Client();
+        pop3Client2.connect("127.0.0.1", server.getProbe(Pop3GuiceProbe.class).getPop3Port());
+        pop3Client2.login(USER, PASSWORD);
+        List<POP3MessageInfo> pop3MessageInfos2 = ImmutableList.copyOf(pop3Client2.listUniqueIdentifiers());
+        assertThat(pop3MessageInfos2).hasSize(1);
+        pop3Client.disconnect();
```

Comment:
you forget to disconnect the pop3Client2 at the end here

Reply:
There is only one POP3 client in this test, right?

Reply:
I see 2. Line 285 `pop3Client` and line 293 `pop3Client2`